### ISSUE TITLE
Karpenter deprovisioning stuck

### ISSUE THREAD

**Issue by ig-leonov** [aws/karpenter#3576]
### Version

__Karpenter Version:__ v0.27.0
__Kubernetes Version:__ v1.23.0

### Expected Behavior

Karpenter should be able to deprovision nodes

### Actual Behavior

After some point in time Karpenter stopped deprovisioning nodes until I manually killed some EC2 instances 

### Steps to Reproduce the Problem

We had a cluster which was scaled to ±1500 nodes.  After some point we noticed that we have much less non-daemonset pods than nodes:
<img width="1086" alt="image" src="https://user-images.githubusercontent.com/71068192/224294987-c2064a75-082c-46d9-bb93-371e2ee1f470.png">

Logs for one of the instances that was running for several hours without any load (the last non-daemonset pod was there at 4:01):
`2023-03-10 05:14:02.817	2023-03-10T10:14:02.817Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 05:04:02.815	2023-03-10T10:04:02.814Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:54:02.814	2023-03-10T09:54:02.814Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:44:02.814	2023-03-10T09:44:02.814Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:34:02.814	2023-03-10T09:34:02.814Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:24:02.813	2023-03-10T09:24:02.813Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:14:02.802	2023-03-10T09:14:02.802Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:04:02.802	2023-03-10T09:04:02.802Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 04:03:37.532	2023-03-10T09:03:37.531Z	DEBUG	controller.node_state	discovered EC2 instance types zonal offerings for subnets	{"commit": "dc3af1a", "node": "NODE_NAME_HERE", "subnet-selector": "{\"Name\":\"SUBNET_NAME\"}"}
2023-03-10 04:03:37.338	2023-03-10T09:03:37.338Z	DEBUG	controller.node_state	discovered subnets	{"commit": "dc3af1a", "node": "NODE_NAME_HERE", "subnets": ["SUBNET_NAME (us-east-1b)"]}
2023-03-10 04:01:36.881	2023-03-10T09:01:36.881Z	INFO	controller.node	added TTL to empty node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:54:02.801	2023-03-10T08:54:02.801Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:44:02.788	2023-03-10T08:44:02.788Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:34:02.779	2023-03-10T08:34:02.779Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:24:02.779	2023-03-10T08:24:02.779Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:14:02.778	2023-03-10T08:14:02.778Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:04:02.772	2023-03-10T08:04:02.772Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 03:00:27.391	2023-03-10T08:00:27.391Z	INFO	controller.node	removed emptiness TTL from node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:59:43.390	2023-03-10T07:59:43.390Z	INFO	controller.node	added TTL to empty node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:54:02.771	2023-03-10T07:54:02.771Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:44:02.771	2023-03-10T07:44:02.771Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:34:02.771	2023-03-10T07:34:02.771Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:24:02.631	2023-03-10T07:24:02.631Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:14:02.631	2023-03-10T07:14:02.631Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 02:04:02.630	2023-03-10T07:04:02.630Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 01:56:57.719	2023-03-10T06:56:57.719Z	INFO	controller.node	removed emptiness TTL from node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 01:56:34.908	2023-03-10T06:56:34.908Z	INFO	controller.node	added TTL to empty node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 01:54:42.763	2023-03-10T06:54:42.763Z	INFO	controller.node	removed emptiness TTL from node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 01:54:20.789	2023-03-10T06:54:20.789Z	INFO	controller.node	added TTL to empty node	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 01:54:02.629	2023-03-10T06:54:02.629Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected 234 of resource pods, but found 50 (21.4% of expected)	{"commit": "dc3af1a", "node": "NODE_NAME_HERE"}
2023-03-10 01:44:02.627	`

After I manually killed a hundred of EC2 instances in AWS console, karpenter was able to deprovision all others,
`INFO	controller.deprovisioning	deprovisioning via emptiness delete, terminating 1404 nodes `
previous "deprovisioning via emptiness delete" message in logs was 8 hours before. So I think something wrong happened with deprovisioning controller

### Resource Specs and Logs

controller.deprovisioning errors:
`2023-03-10 05:17:49.724	2023-03-10T10:17:49.724Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 05:08:15.300	2023-03-10T10:08:15.300Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:58:41.004	2023-03-10T09:58:41.004Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:49:06.837	2023-03-10T09:49:06.837Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:39:32.537	2023-03-10T09:39:32.537Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:29:58.222	2023-03-10T09:29:58.222Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:20:23.976	2023-03-10T09:20:23.975Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:10:49.684	2023-03-10T09:10:49.683Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 04:01:15.143	2023-03-10T09:01:15.143Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 03:51:40.737	2023-03-10T08:51:40.737Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 03:42:06.200	2023-03-10T08:42:06.200Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 03:32:31.850	2023-03-10T08:32:31.850Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 03:22:57.571	2023-03-10T08:22:57.571Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 03:13:23.128	2023-03-10T08:13:23.127Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 03:03:48.789	2023-03-10T08:03:48.789Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 02:54:14.576	2023-03-10T07:54:14.576Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 02:44:40.216	2023-03-10T07:44:40.216Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 02:35:05.854	2023-03-10T07:35:05.854Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 02:25:31.134	2023-03-10T07:25:31.134Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 02:15:56.254	2023-03-10T07:15:56.254Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 02:06:21.423	2023-03-10T07:06:21.423Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 01:56:47.023	2023-03-10T06:56:47.023Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 01:47:12.816	2023-03-10T06:47:12.816Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 01:37:38.596	2023-03-10T06:37:38.596Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 01:28:04.231	2023-03-10T06:28:04.231Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 01:18:29.948	2023-03-10T06:18:29.948Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 01:08:55.575	2023-03-10T06:08:55.575Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:59:21.314	2023-03-10T05:59:21.314Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:49:47.125	2023-03-10T05:49:47.125Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:40:12.789	2023-03-10T05:40:12.789Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:30:38.533	2023-03-10T05:30:38.533Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:21:04.276	2023-03-10T05:21:04.276Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:11:29.816	2023-03-10T05:11:29.816Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-10 00:01:55.490	2023-03-10T05:01:55.490Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 23:52:21.221	2023-03-10T04:52:21.221Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 23:42:46.937	2023-03-10T04:42:46.937Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 23:33:12.571	2023-03-10T04:33:12.571Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 23:23:38.336	2023-03-10T04:23:38.336Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 23:14:04.112	2023-03-10T04:14:04.112Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 23:04:29.817	2023-03-10T04:04:29.817Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 22:54:55.456	2023-03-10T03:54:55.456Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 22:45:21.098	2023-03-10T03:45:21.098Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 22:35:46.885	2023-03-10T03:35:46.885Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 22:26:12.659	2023-03-10T03:26:12.659Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 22:16:38.352	2023-03-10T03:16:38.352Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 22:07:03.827	2023-03-10T03:07:03.827Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 21:57:29.392	2023-03-10T02:57:29.392Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 21:47:54.965	2023-03-10T02:47:54.965Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 21:38:20.641	2023-03-10T02:38:20.641Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}
2023-03-09 21:28:46.193	2023-03-10T02:28:46.193Z	ERROR	controller.deprovisioning	Waiting on node deletion, expected node to be not found	{"commit": "dc3af1a"}`

Another error in logs which might be related:
`2023-03-10T07:31:38.322Z	ERROR	controller.pod_state	inconsistent state error adding volume, PersistentVolumeClaim "blahblah" not found, please file an issue	{"commit": "dc3af1a", "pod": "blahblahblahblah"}`  

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tzneal** [comment]
Can you attach the full set of Karpenter logs?

**Comment by tzneal** [comment]
Just the Karpenter controller logs, I wanted to see what it was doing prior to this point.

**Comment by spring1843** [comment]
@ig-leonov can you please also share your provisioner or important aspects of it so that we can try to reproduce?

**Comment by bwagner5** [comment]
Unable to reproduce deprovisioning getting stuck so far.

**Comment by tzneal** [comment]
This should be resolved in v0.27.2, please re-open or file a new issue if you still experience this.

**Comment by shay-ul** [comment]
@tzneal 

Hey @tzneal, we are experiencing a similar issue with Karpenter 0.27.2. Very slow to non-existent deprovisioning on nodes that are near empty. Restarting the Karpenter deployment immediately initiates the process (controller.deprovisioning) and nodes start to deprovision quickly.
On some nodes we see the "not all pods would schedule" Events before we restart the Karpenter deployment, but after the restart those nodes are immediately deleted.
This is very costly for us since we are running at scale with peak hours. in peak hours we run ~200 nodes which we expect to decrease over 4 hours to around ~30 nodes. This means the consolidation feature is critical for us in terms of cost.
Any idea how we can troubleshoot this?
Thanks

**Comment by tzneal** [comment]
@shay-ul Can you create a new issue and include your Karpenter controller logs?
